### PR TITLE
Create app

### LINKED ISSUE #4: Build sample app
`mycoind` is obsolete.

Let's do a fresh start and add a simple application that combines some weave modules together, along with nice comments to make it easy to customize. This should be referenced in the documentation.

It will serve the purpose of mycoind, but can be simplified.

### PR DESCRIPTION
**Pull request by orkunkl**
Closes #4 
This is basically a copy-paste from the tutorial. If you think if there is an unnecessary chain decorator, let me know

### CODE REVIEW

File: app/app.go
Comment (ethanfrey) [inline app/app.go]: Replace this TODO with the multisig and cash router at least.

File: app/app.go
Comment (ethanfrey) [inline app/app.go]: msgfee can probably be removed from base app.
And the cash.StaticFeeDecorator can be used above (this is the simplest case when the product fee doesn't depend on the output from the Handler)

File: app/app.go
Comment (ethanfrey) [inline app/app.go:71]: I would add cash here, as this is basic to most apps.
And `msgfee` if you don't remove the decorator above.

File: app/codec.proto
Comment (ethanfrey) [inline app/codec.proto:34]: You can also add the create/update multisig msgs here

File: app/codec.proto
Comment (ethanfrey) [inline app/codec.proto]: If you add these, please update the Router and QueryHandler set in the app to include these

File: app/init.go
Comment (ethanfrey) [inline app/init.go]: Please update this to be in-line with the router/query/etc.

There are a few places that need to be kept in sync - good to add this to the docs.

When adding a package:
- `Stack` if it has a Decorator
- `Router` if it has a Handler
- `QueryRouter` if it has any Buckets
- `DecorateApp`/init if it has any Initializer.

Best to write down a list of all modules to use first, then make them all the same. As discussed above, I think the minimal set is:

* cash
* sigs
* multisig
* currency
* migration (as a runtime dependency of above modules)

`validators` is a nice to have to run a PoA testnet and good to have as a demo.

--

`msgfee` is a bit more complex and requires more complex Fee handling. I would not use this for default app.

File: x/custom/codec.proto
Comment (ethanfrey) [inline x/custom/codec.proto:5]: This didn't work the other way?
I actually thought we wanted short paths for references in the same repo, and full for outside.

But if needed for protoc, then do it like this.

**Review by ethanfrey — COMMENTED** [review]
Nice start. Made some comments to clean up app.
Clearly docs need to exist to explain this. Thanks for figuring this out without proper docs.

**Comment by ethanfrey** [comment]
About protoc... these are key lines in prototool.yaml.

```
 includes:
    - .
    - /usr/include
    - spec
    - spec/github.com/iov-one/weave
```

`.` means that `import "x/custom/codec.go"` will work inside this app (recommended to avoid GOPATH issues)

`spec` means that we can `import "github.com/iov-one/weave/coins/codec.proto"` and it will use the protobuf file from our spec dir (which must be kept up to date with the weave repo).

`spec/github.com/iov-one/weave` is needed as if we import eg. `github.com/iov-one/weave/x/cash/codec.proto`, it imports `coins/codec.proto` using relative imports to it's project. When we then import this via spec, the relative import doesn't work without this second line.

That is - customapp imports modules from same repo with short path. customapp imports weave protobuf from spec with full path. weave protobuf inside spec imports other weave protobuf with short path.

Does this make sense? It was how I got it to work once (how it still does). But I really should have documented this better. Sorry.

File: x/custom/codec.proto
Comment (orkunkl) [reply x/custom/codec.proto:5]: I tried to make it work as it was but no chance :disappointed: The only thing worked is to make `x/codec` imports refer to weave codec without the prefix.

Comment (ethanfrey) [reply x/custom/codec.proto:5]: Okay, then we just document that. Maybe someone (maybe me?) has time to look into it. Otherwise we just explain how to work with this.

File: app/app.go
Comment (orkunkl) [reply app/app.go:71]: I thought we should keep weave-starter-kit simple as possible. So deleted `msgfee`

File: app/codec.proto
Comment (orkunkl) [reply app/codec.proto]: I still need to implement handler and buckets. Lets keep this PR on until I finish and merge them

File: x/custom/codec.proto
Comment (orkunkl) [reply x/custom/codec.proto:5]: I would appreciate it if somebody takes a look at it. I spent quite a time on the issue and this is the solution I come up with  ¯\_(ツ)_/¯

**Comment by orkunkl** [comment]
Blocked by https://github.com/iov-one/weave-starter-kit/pull/31

File: app/app.go
Comment (husio) [inline app/app.go]: I am not sure why this comment is here. I think if at all, it should be part of general application documentation or package level comment. This place seems quite random for this listing.

File: app/app.go
Comment (husio) [inline app/app.go]: Error string should not be capitalized https://github.com/golang/go/wiki/CodeReviewComments#error-strings
Maybe it is worth using `weave/errors` package to produce a better error instance?

File: app/app.go
Comment (husio) [inline app/app.go]: I think it is better to wrap error with additional information. Such context information is useful during debugging. For example:
```go
return app.BaseApp{}, errors.Wrap(err, "cannot create database instance")
```

File: app/codec.proto
Comment (husio) [inline app/codec.proto:26]: Is the above comment true? This attribute will generate `GetCashFees` method. :thinking:

File: app/codec.proto
Comment (husio) [inline app/codec.proto]: This attribute is called `sum` not `msg` :police_car:

File: app/init.go
Comment (husio) [inline app/init.go]: I think it is worth adding a comment what collector address is (to whom it belongs to?) and why it is hardcoded here.

File: app/init.go
Comment (husio) [inline app/init.go]: I think a TODO does not belong in a good codebase. Either implement it or leave a good comment explaining *why* below functionality works in this certain way and what was the reason for choosing this implementation.

Leaving TODO like this will give an impression that the code is not finished and lacking, therefore cannot be trusted. No one will address this comment as well.

File: app/tx.go
Comment (husio) [inline app/tx.go]: I think a much better comment explaining the functionality would be

> GetMsg returns a single message instance that is represented by this transaction.

**Review by husio — APPROVED** [review]
If you want to make this crips, please update the comments to follow https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

File: app/app.go
Comment (orkunkl) [reply app/app.go]: Nice advise :+1: I really enjoy writing go because only language itself covers and enforces structuring and documenting your code.

File: app/codec.proto
Comment (orkunkl) [reply app/codec.proto:26]: https://github.com/iov-one/weave/blob/master/x/cash/codec.pb.go#L220-L225 Auto-generation does not cover the concept though.

File: app/init.go
Comment (orkunkl) [reply app/init.go]: Nice point. I will take `bnsd/init` as an example because `mycoind/init`seems too much outdated

File: app/codec.proto
Comment (alpe) [reply app/codec.proto:26]: It is not necessary to add comments for autogeneration. Just remove it to not create confusion with all the other fields where it is not added.

**Review by alpe — COMMENTED** [review]
Only had a very brief look:

- Please run `go fmt $(go list ./... | grep -v "/vendor/")`
- `make protoc` fails on my box 
```
<input>:1:1:Import "x/custom/codec.proto" was not found.
app/codec.proto:29:5:"custom.CreateTimedStateMsg" is not defined.
app/codec.proto:30:5:"custom.CreateStateMsg" is not defined.
```
* `make test` fails on my box:
```
app/app.go:23:2: unknown import path "github.com/iov-one/weave-starter-kit/x/custom": import lookup disabled by -mod=readonly
```

**Comment by orkunkl** [comment]
@alpe really weird you are having errors on `make protoc`. I don't have any. Do you have any guess why you could be getting errors?

**Comment by orkunkl** [comment]
Waiting for response from https://github.com/gogo/protobuf/issues/601

**Comment by orkunkl** [comment]
I have fixed harmless `protoc` bug in https://github.com/iov-one/weave/pull/953. Changes will be applied when upgraded to weave v0.21.0

**Pull request merged**
